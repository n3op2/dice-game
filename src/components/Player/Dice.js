import React from 'react';

import up from '/Users/paulius/dice-game/assets/dice-one-solid.svg';
import back from '/Users/paulius/dice-game/assets/dice-two-solid.svg';
import left from '/Users/paulius/dice-game/assets/dice-three-solid.svg';
import right from '/Users/paulius/dice-game/assets/dice-four-solid.svg';
import front from '/Users/paulius/dice-game/assets/dice-five-solid.svg';
import down from '/Users/paulius/dice-game/assets/dice-six-solid.svg';

const sides = {
    1: { name: 'top', img: up },
    2: { name: 'back', img: back },
    3: { name: 'left', img: left },
    4: { name: 'right', img: right },
    5: { name: 'front', img: front },
    6: { name: 'bottom', img: down },
};

const Dice = ({ rolling, val }) => {
    const [side, setSide] = React.useState({ img: up, name: 'up' });

    React.useEffect(() => {
        if (val === 0) return null;
        setSide(sides[val]);
    }, [rolling, val]);

    return (
        <img
            name={side.name}
            className={rolling ? 'dice-rolling' : 'dice'}
            src={side.img}
        />
    );
};

export default Dice;
