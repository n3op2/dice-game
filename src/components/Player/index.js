import React from 'react';

import Dice from './Dice';
import { AppContext, Container, Text } from '../shared';

const rnd = () => Math.ceil(Math.random() * 6);

const Player = (props) => {
    const { update, playing, ...ctx } = React.useContext(AppContext);
    const { name } = props;
    const self = { ...ctx[name], ...props };

    React.useEffect(() => {
        if (playing && !self.rolling)
            update({ [name]: { ...self, rolling: true } });
    }, [playing]);

    React.useEffect(() => {
        if (playing && self.rolling) {
            // await new Promise(r => setTimeout(r, Math.random() * 1000 + 100))
            update({ [name]: { ...self, val: rnd(), rolling: false } });
        }
    }, [self.rolling]);

    return (
        <Container style={{ width: 200 }}>
            <Text>name: {name}</Text>
            <Dice rolling={self.rolling} val={self.val} />
            <Text>score: {self.score.toString()}</Text>
        </Container>
    );
};

export default Player;
