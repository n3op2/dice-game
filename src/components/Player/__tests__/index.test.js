import React from 'react';
import { mount } from 'enzyme';
import Player from '../';

const mockPlayer = {
    score: 0,
    val: 0,
    name: 'test',
    rolling: false,
};

describe('Player Component', () => {
    let wrapper;
    let mockContext = {
        update: jest.fn(),
        playing: true,
        log: jest.fn(),
        test: mockPlayer,
    };

    beforeEach(() => {
        jest.spyOn(global.Math, 'random').mockReturnValue(1); // will be multiplied by 6
        React.useEffect = jest.spyOn(React, 'useEffect');
        React.useContext = jest.fn().mockReturnValue(mockContext);
        wrapper = mount(<Player {...mockPlayer} />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.spyOn(global.Math, 'random').mockRestore();
    });

    test('it renders a component', () => {
        expect(wrapper.props()).toEqual(mockPlayer);
        expect(wrapper.find('Dice').props()).toEqual({
            val: 0,
            rolling: false,
        });
        expect(wrapper).toMatchSnapshot();
    });

    describe('after the has started', () => {
        test('sets rolling to true', () => {
            expect(mockContext.update).toHaveBeenCalledWith({
                [mockPlayer.name]: expect.objectContaining({
                    rolling: true,
                    val: 0,
                }),
            });
        });

        test('sets rolling to true and update dice', () => {
            wrapper.setProps({ rolling: true });
            expect(wrapper.find('Dice').prop('rolling')).toBe(true);
            expect(wrapper.find('Dice').find('img').prop('className')).toEqual(
                'dice-rolling'
            );
        });

        test('updates val and sets rolling to false', async () => {
            wrapper.setProps({ rolling: true });
            await new Promise(r => setTimeout(r, 2000))
            expect(mockContext.update).toHaveBeenLastCalledWith({
                test: {
                    ...mockPlayer,
                    rolling: false,
                    val: 6,
                },
            });
        });

        test('re-renders dice component', () => {
            wrapper.setProps({ rolling: true, val: 6 });
            expect(wrapper.find('Dice')).toMatchSnapshot();
        });
    });
});
