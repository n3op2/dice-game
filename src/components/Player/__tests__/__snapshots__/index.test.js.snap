// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Component after the has started re-renders dice component 1`] = `
<Dice
  rolling={true}
  val={6}
>
  <img
    className="dice-rolling"
    name="up"
    src={Object {}}
  />
</Dice>
`;

exports[`Player Component it renders a component 1`] = `
<Player
  name="test"
  rolling={false}
  score={0}
  val={0}
>
  <Container
    style={
      Object {
        "width": 200,
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "marginBottom": 20,
          "marginTop": 20,
          "width": 200,
        }
      }
    >
      <Text>
        <div
          className="text"
        >
          name: 
          test
        </div>
      </Text>
      <Dice
        rolling={false}
        val={0}
      >
        <img
          className="dice"
          name="up"
          src={Object {}}
        />
      </Dice>
      <Text>
        <div
          className="text"
        >
          score: 
          0
        </div>
      </Text>
    </div>
  </Container>
</Player>
`;
